pkg/test/utils/utils.go

```diff
@@ -0,0 +1,171 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+// Start the envtest environment
+func StartTestEnv(scheme *runtime.Scheme, testEnv *envtest.Environment) client.Client {
+	cfg, err := testEnv.Start()
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error creating the envtest environment during test setup: %v", err))
+	kubeconfigPath := CreateKubeconfigFileForRestConfig(*cfg)
+	os.Setenv("KUBECONFIG", kubeconfigPath)
+	k8sClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error creating the client during test setup: %v", err))
+	Expect(k8sClient).NotTo(BeNil())
+
+	return k8sClient
+}
+
+// Stop the envtest environment
+func StopEnvTest(envTest *envtest.Environment) {
+	if envTest != nil {
+		err := envTest.Stop()
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Failed to stop envTest: %v", err))
+	}
+}
+
+// Create a Kubeconfig from the given rest config.
+func CreateKubeconfigFileForRestConfig(restConfig rest.Config) string {
+	clusters := make(map[string]*clientcmdapi.Cluster)
+	clusters["default-cluster"] = &clientcmdapi.Cluster{
+		Server:                   restConfig.Host,
+		CertificateAuthorityData: restConfig.CAData,
+	}
+	contexts := make(map[string]*clientcmdapi.Context)
+	contexts["default-context"] = &clientcmdapi.Context{
+		Cluster:  "default-cluster",
+		AuthInfo: "default-user",
+	}
+	authinfos := make(map[string]*clientcmdapi.AuthInfo)
+	authinfos["default-user"] = &clientcmdapi.AuthInfo{
+		ClientCertificateData: restConfig.CertData,
+		ClientKeyData:         restConfig.KeyData,
+	}
+	clientConfig := clientcmdapi.Config{
+		Kind:           "Config",
+		APIVersion:     "v1",
+		Clusters:       clusters,
+		Contexts:       contexts,
+		CurrentContext: "default-context",
+		AuthInfos:      authinfos,
+	}
+	kubeConfigFile, _ := os.CreateTemp("", "kubeconfig")
+	_ = clientcmd.WriteToFile(clientConfig, kubeConfigFile.Name())
+	return kubeConfigFile.Name()
+}
+
+// Build the workspace-manager binary.
+// mainPath is the path to the main module.
+func BuildWorkspaceManager(mainPath string) string {
+	out := os.TempDir()
+	binPath := filepath.Join(out, "workspace-manager", "manager")
+	buildCmd := exec.Command("go", "build", "-o", binPath, mainPath)
+	buildLog, err := buildCmd.CombinedOutput()
+	if err != nil {
```

Comment:
Do we need this if-statement?

Reply:
Same for another statement below, the `if` + `expect` seems redundant.

Reply:
it's needed since `err.Error()` is called for getting the error message. Without the `if`, if the error is nil line 83 will break with nil pointer exception.